clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -5402,6 +5407,10 @@ LangAS CodeGenModule::GetGlobalConstantAddressSpace() const {
     // UniformConstant storage class is not viable as pointers to it may not be
     // casted to Generic pointers which are used to model HIP's "flat" pointers.
     return LangAS::cuda_device;
+  if (LangOpts.OpenMPIsTargetDevice && getTriple().isSPIRV())
+    // OpenMP SPIR-V global constants should map to UniformConstant, different
+    // from the HIPSPV case above.
+    return LangAS::opencl_constant;
```

Comment:
I believe that whilst this makes sense you might run into obnoxious issues where valid source ends up generating a SPIR-V invalid cast to/from generic/constant, which will fail in the translator / fail SPIR-V validation (HIP code runs into this). I *think* we need to relax this restriction at least in the translator / BE, or potentially extend SPIR-V itself in this direction.

Reply:
Yeah I hit something similar when trying to use `addrspace(4)` for generic. For constants it seems to be working okay for now, hopefully it's okay with you if I try this for now and then extend the solution if a problem comes up.